service/history/workflow/update/update.go

```diff
@@ -159,31 +163,86 @@ func newCompleted(
 // the acceptance occurs or the provided context.Context expires.
 // It is safe to call this method outside of workflow lock.
-func (u *Update) WaitAccepted(ctx context.Context) (*updatepb.Outcome, error) {
+func (u *Update) WaitAccepted(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
 	if u.outcome.Ready() {
-		// being complete implies being accepted, return the completed outcome
+		// Being complete implies being accepted; return the completed outcome
 		// here because we can.
-		return u.outcome.Get(ctx)
+		return u.WaitOutcome(ctx)
 	}
 	fail, err := u.accepted.Get(ctx)
 	if err != nil {
-		return nil, err
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, err
 	}
 	if fail != nil {
-		return &updatepb.Outcome{
+		outcome := &updatepb.Outcome{
 			Value: &updatepb.Outcome_Failure{Failure: fail},
-		}, nil
+		}
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, outcome, nil
+	}
+	return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED, nil, nil
+}
+
+// WaitLifecycleStage waits until the Update has reached at least `waitStage` or
+// a timeout. If the Update reaches `waitStage` with no timeout, the most
+// advanced stage known to have been reached is returned, along with the outcome
+// if any. If there is a timeout due to the supplied soft timeout, then
+// unspecified stage and nil outcome are returned, without an error. If there is
+// a timeout due to context deadline expiry, then the error is returned as usual.
+func (u *Update) WaitLifecycleStage(ctx context.Context, waitStage enumspb.UpdateWorkflowExecutionLifecycleStage, softTimeout time.Duration) (stage enumspb.UpdateWorkflowExecutionLifecycleStage, outcome *updatepb.Outcome, err error) {
+	stage = enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED
+	switch waitStage {
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED:
+		return u.waitLifecycleStage(ctx, u.WaitAccepted, softTimeout)
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED:
+		return u.waitLifecycleStage(ctx, u.WaitOutcome, softTimeout)
+	default:
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, serviceerror.NewUnimplemented(
+			fmt.Sprintf("%v is not implemented", waitStage))
+	}
+}
+
+func (u *Update) waitLifecycleStage(ctx context.Context, waitFn func(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error), softTimeout time.Duration) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
+
+	type result struct {
+		stage   enumspb.UpdateWorkflowExecutionLifecycleStage
+		outcome *updatepb.Outcome
+		err     error
+	}
+	ch := make(chan result, 1)
+
+	go func() {
+		ctx, cancel := context.WithTimeout(ctx, softTimeout)
+		defer cancel()
+		stage, outcome, err := waitFn(ctx)
+		if err != nil && common.IsContextDeadlineExceededErr(err) {
```

Comment:
How do you distinguish between the original context timeout and this "softTimeout" (e.g. the soft timeout is greater than the context timeout)?
There could be a race here AFAICT where the context deadline exceeds and you incorrectly return a nil error.

Reply:
Thanks, good catch. When I wrote that I mistakenly thought that `WithTimeout(ctx, t)` overrides the deadline in `ctx` unconditionally. I've changed it to compare against `context.Background()` (as you had it).